### PR TITLE
Us012 add role

### PR DESCRIPTION
**Pull request by jackpollinsK**

### CODE REVIEW

File: views/snippets/nav.html [hunk 1]
```diff
@@ -2,15 +2,15 @@
     {href: '/', id: 'Kainos Jobs Homepage', caption: 'Home'},
     {href: '/jobRoles', id: 'Job Roles', caption: 'Jobs'},
```
Comment (ThomasS-kainos) [inline views/snippets/nav.html]: Changed in main make sure to merge

**Review by ThomasS-kainos — CHANGES_REQUESTED** [review]

File: views/snippets/nav.html (hunk 1 above)
Comment (ThomasS-kainos) [reply views/snippets/nav.html:3]: /AIJobSearch

**Review by ThomasS-kainos — DISMISSED** [review]

**Review by AdamB-Kainos — DISMISSED** [review]

File: src/main/controllers/JobRoleController.ts [hunk 2]
```diff
@@ -33,4 +35,69 @@ export const deleteJobRole = async (req: express.Request, res: express.Response)
...
\ No newline at end of file
+}
+
+export const getRoleForm = async (req: express.Request, res: express.Response): Promise<void> => {
+    try {
+        const [bands, locations, capabilities] = await Promise.all([
+            getBands(req.session.token),
+            getLocations(req.session.token),
+            getCapabilities(req.session.token)
+        ]);
```
Comment (shaunganley) [inline src/main/controllers/JobRoleController.ts:45]: love the use of promise.all here, nice job

File: src/main/controllers/JobRoleController.ts [hunk 3]
```diff
@@ -33,4 +35,69 @@ export const deleteJobRole = async (req: express.Request, res: express.Response)
...
+        res.locals.errormessage = e.message;
+        res.locals.pageName = "An Error Occurred";
+        res.render('pages/errorPage.html');
+    }
+}
+
+export const postRoleForm = async (req: express.Request, res: express.Response): Promise<void> => {
+    try {
+        const jobRoleRequest: JobRoleRequest = req.body;
+        const errors = validateJobRoleRequest(jobRoleRequest);
```
Comment (shaunganley) [inline src/main/controllers/JobRoleController.ts]: validation should happen in the service

File: src/main/controllers/JobRoleController.ts [hunk 4]
```diff
@@ -33,4 +35,69 @@ export const deleteJobRole = async (req: express.Request, res: express.Response)
...
+                errorMessages: errors,
+                roleName: jobRoleRequest.roleName,
+                description: jobRoleRequest.description,
+                responsibilities: jobRoleRequest.responsibilities,
+                jobSpec: jobRoleRequest.jobSpec,
+                closingDate: jobRoleRequest.closingDate,
+                openPositions: jobRoleRequest.openPositions,
+                selectedLocation: jobRoleRequest.location,
+                selectedBand: jobRoleRequest.band,
+                selectedCapability: jobRoleRequest.capability,
```
Comment (shaunganley) [inline src/main/controllers/JobRoleController.ts]: can't you just pass back the whole request object?

File: src/main/controllers/JobRoleController.ts [hunk 5]
```diff
@@ -33,4 +35,69 @@ export const deleteJobRole = async (req: express.Request, res: express.Response)
...
+            });
+        } else {
+            await createRole(jobRoleRequest, req.session.token);
+            res.redirect('/');
```
Comment (shaunganley) [inline src/main/controllers/JobRoleController.ts]: redirect to the job role details screen

File: src/test/unit/controller/JobRoleControllerTests.ts [hunk 6]
```diff
@@ -181,6 +183,176 @@ describe('JobRoleController', function () {
...
+                    closingDate: new Date(),
+                    description: 'Description',
+                    responsibilities: 'Responsibilities',
+                    jobSpec: 'Job Spec',
+                    openPositions: 5
+                };
+                const req = { body: jobRoleRequest, session: { token: validAdminJwtToken } };
+                const res = { redirect: sinon.stub() };
+    
+                sinon.stub(JobRoleService, 'createRole').resolves();
```
Comment (shaunganley) [inline src/test/unit/controller/JobRoleControllerTests.ts]: this should resolve an id which we can use to redirect to details page

**Review by shaunganley — CHANGES_REQUESTED** [review]
needs validator tests

**Review by AdamB-Kainos — APPROVED** [review]

File: src/main/controllers/JobRoleController.ts [hunk 7]
```diff
@@ -61,43 +60,38 @@ export const getRoleForm = async (req: express.Request, res: express.Response):
...
+        try {
             const [bands, locations, capabilities] = await Promise.all([
                 getBands(req.session.token),
                 getLocations(req.session.token),
                 getCapabilities(req.session.token)
             ]);
 
+            const errorMessages = e.message === 'Invalid Details'
+                ? ['Invalid Details']
+                : [e.message];
```
Comment (shaunganley) [inline src/main/controllers/JobRoleController.ts:79]: I don't think this actually does anything

**Review by shaunganley — APPROVED** [review]

**Pull request merged**